### PR TITLE
Provide interface to collect stream events sizes

### PR DESCRIPTION
**Pull request by jguery**
Rather than performing the metrics logic in the ``StreamConcurrentWriter``, we create a ``StreamMetricsCollector`` defined in the ``StreamHandler`` than can add to the logic of metrics collection. In future PRs, we will use that interface to collect the size of events ingested by a stream, to perform aggregation logic in each stream handler.

### CODE REVIEW

**Comment by jguery** [comment]
Build is running at https://builds.cask.co/browse/CDAP-DUT547.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java [hunk 1]
```diff
@@ -462,7 +461,7 @@ boolean tryAppendFile(StreamConfig streamConfig, Location eventFile, Location in
         writerFlag.set(false);
       }
 
-      emitMetrics(fileSize, eventCount);
+      metricsCollector.emitMetrics(metricsCollector, fileSize, eventCount);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriter.java]: Do we need to pass in `metricsCollector` to itself?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Emit stream metrics.
+   *
+   * @param metricsCollector child {@link MetricsCollector} of the stream to publish metrics about
+   * @param bytesWritten number of bytes written to the string
+   * @param eventsWritten number of events written to the string
+   */
+  public abstract void emitMetrics(MetricsCollector metricsCollector, long bytesWritten, long eventsWritten);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: Looks like `delegate` is enough to emit metrics here, right? No need for `metricsCollector`. Or am I missing something?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: When need a metrics collector if there is `delegate`?

**Comment by poornachandra** [comment]
@jguery Done one pass, please address the comments. I'll continue after that.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java (hunk 2 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: Right, I got confused with the level of indirection. Will remove the ``metricsCollector`` parameter.

**Comment by jguery** [comment]
@poornachandra Thanks, I addressed your comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  private final MetricsCollector delegate;
+
+  protected StreamMetricsCollector(MetricsCollector delegate) {
+    this.delegate = delegate;
+  }
+
+  /**
+   * Emit stream metrics.
+   *
+   * @param bytesWritten number of bytes written to the string
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: number of bytes written to the stream?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   * @param eventsWritten number of events written to the string
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: number of bytes written to the stream?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriterTestBase.java [hunk 5]
```diff
@@ -320,26 +320,35 @@ public StreamConfig getConfig(String streamName) throws IOException {
...
+        public MetricsCollector childCollector(String tagName, String tagValue) {
+          return this;
+        }
+      });
     }
 
     @Override
-    public MetricsCollector childCollector(String tagName, String tagValue) {
-      return this;
+    public void emitMetrics(String streamName, long bytesWritten, long eventsWritten) {
```
Comment (chtyim) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/service/ConcurrentStreamWriterTestBase.java]: Why need stream name?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Emit stream metrics.
+   *
+   * @param streamName name of the stream to emit metrics for
+   * @param bytesWritten number of bytes written to the stream
+   * @param eventsWritten number of events written to the stream
+   */
+  public abstract void emitMetrics(String streamName, long bytesWritten, long eventsWritten);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamMetricsCollector.java]: If stream name cannot change then we should probably pass it in constructor.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 7]
```diff
@@ -112,7 +112,19 @@ public StreamHandler(CConfiguration cConf, Authenticator authenticator,
     this.exploreFacade = exploreFacade;
     this.exploreEnabled = cConf.getBoolean(Constants.Explore.EXPLORE_ENABLED);
 
-    this.metricsCollector = metricsCollectionService.getCollector(MetricsScope.SYSTEM, getMetricsContext());
+    MetricsCollector collector = metricsCollectionService.getCollector(MetricsScope.SYSTEM, getMetricsContext());
+    this.metricsCollector = new StreamMetricsCollector(collector) {
+      @Override
+      public void emitMetrics(String streamName, long bytesWritten, long eventsWritten) {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: `streamName` is not used here?

**Comment by jguery** [comment]
@poornachandra, @chtyim, I addressed your comments.

**Comment by poornachandra** [comment]
LGTM :+1:

**Pull request merged**